okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemoteExceptionResponseHandler.java

```diff
@@ -27,54 +27,50 @@
 import java.io.Reader;
 import java.nio.charset.StandardCharsets;
 import java.util.Collection;
-import javax.annotation.CheckForNull;
+import java.util.Optional;
 import javax.ws.rs.core.MediaType;
-import javax.ws.rs.core.Response;
+import okhttp3.Response;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-/**
- * Attempts to convert a HTTP {@link Response} body as a JSON representation of a {@link SerializableError} and
- * re-create the original (or close-to) exception including exception type, message, and stacktrace. Creates {@link
- * RuntimeException} if the body cannot be interpreted as a {@link SerializableError}, or if the exception otherwise
- * fails to get re-created.
- */
-public final class SerializableErrorToExceptionConverter {
-
-    private SerializableErrorToExceptionConverter() {}
-
-    private static final Logger log = LoggerFactory.getLogger(SerializableErrorToExceptionConverter.class);
+enum RemoteExceptionResponseHandler implements ResponseHandler<RemoteException> {
+    INSTANCE;
 
+    private static final Logger log = LoggerFactory.getLogger(RemoteExceptionResponseHandler.class);
     private static final ObjectMapper MAPPER = ObjectMappers.newClientObjectMapper();
 
-    public static RuntimeException getException(Collection<String> contentTypes, int status,
-            @CheckForNull InputStream body) {
-        if (body == null) {
-            return new RuntimeException(Integer.toString(status));
+    @Override
+    public Optional<RemoteException> handle(Response response) {
+        if (response.body() == null || response.body().byteStream() == null
+                || response.isSuccessful()
+                || response.code() == 101) {
+            return Optional.empty();
         }
 
-        String bodyAsString = readBodyAsString(body);
+        Collection<String> contentTypes = response.headers("Content-Type");
         if (contentTypes.contains(MediaType.APPLICATION_JSON)) {
+            final String body;
             try {
-                SerializableError serializableError = MAPPER.readValue(bodyAsString, SerializableError.class);
-                return new RemoteException(serializableError, status);
+                body = toString(response.body().byteStream());
+            } catch (IOException e) {
+                return Optional.empty();
+            }
+            try {
+                SerializableError serializableError = MAPPER.readValue(body, SerializableError.class);
+                return Optional.of(new RemoteException(serializableError, response.code()));
             } catch (Exception e) {
-                String message = String.format(
+                log.warn("Failed to deserialize JSON: {}", String.format(
```

Comment:
This log.warn is a bit strange... the String.format gives us a double message.  Also if we used normal `{}` then you could probably safeArg the response.code() at least